validation/manifest.go

```diff
@@ -33,79 +132,111 @@ func ValidateManifestWithGroupSpecs(m *manifest.Manifest, gspecs []*dtypes.Group
-	return nil
+	rgroups := make([]types.ResourceGroup, 0, len(dgroups))
+	for _, dgroup := range dgroups {
+		rgroups = append(rgroups, dgroup)
+
+	}
+
+	return validateManifestDeploymentGroups(m.GetGroups(), rgroups)
 }
 
 func validateManifestDeploymentGroups(mgroups []manifest.Group, dgroups []types.ResourceGroup) error {
-
 	if len(mgroups) != len(dgroups) {
 		return errors.Errorf("invalid manifest: group count mismatch (%v != %v)", len(mgroups), len(dgroups))
 	}
 
-mainloop:
+	dgroupByName := make(map[string]types.ResourceGroup)
+
+	for _, dgroup := range dgroups {
+		dgroupByName[dgroup.GetName()] = dgroup
+	}
+
 	for _, mgroup := range mgroups {
-		for _, dgroup := range dgroups {
-			if mgroup.GetName() != dgroup.GetName() {
-				continue
-			}
-			if err := validateManifestDeploymentGroup(mgroup, dgroup); err != nil {
-				return err
-			}
-			continue mainloop
+		dgroup, dgroupExists := dgroupByName[mgroup.GetName()]
+
+		if !dgroupExists {
+			return errors.Errorf("invalid manifest: unknown deployment group ('%v')", mgroup.GetName())
+		}
+
+		if err := validateManifestDeploymentGroup(mgroup, dgroup); err != nil {
+			return err
 		}
-		return errors.Errorf("invalid manifest: unknown deployment group ('%v')", mgroup.GetName())
 	}
+
 	return nil
 }
 
-func validateManifestDeploymentGroup(mgroup types.ResourceGroup, dgroup types.ResourceGroup) error {
+func validateManifestDeploymentGroup(mgroup manifest.Group, dgroup types.ResourceGroup) error {
 	mlist := make([]types.Resources, len(mgroup.GetResources()))
 	copy(mlist, mgroup.GetResources())
 
-mainloop:
-	for _, drec := range dgroup.GetResources() {
+	endpointsCountForDeploymentGroup := 0
 
+	// Iterate over all deployment groups
+deploymentGroupLoop:
+	for _, drec := range dgroup.GetResources() {
+		endpointsCountForDeploymentGroup += len(drec.Resources.Endpoints)
+		// Find a matching manifest group
 		for idx := range mlist {
 			mrec := mlist[idx]
 
+			// Check that this manifest group is not yet exhausted
 			if mrec.Count == 0 {
 				continue
 			}
 
-			if !drec.Resources.Equals(mrec.Resources) {
+			if !drec.Resources.CPU.Equal(mrec.Resources.CPU) ||
+				!drec.Resources.Memory.Equal(mrec.Resources.Memory) ||
+				!drec.Resources.Storage.Equal(mrec.Resources.Storage) {
 				continue
 			}
 
-			if mrec.Count > drec.Count {
+			// If the manifest group contains more resources than the deploynent group, then
+			// fulfill the deployment group entirely
+			if mrec.Count >= drec.Count {
 				mrec.Count -= drec.Count
 				drec.Count = 0
 			} else {
+				// Partially fulfill the deployment group since the manifest group contains less
 				drec.Count -= mrec.Count
 				mrec.Count = 0
 			}
 
+			// Update the value stored in the list
 			mlist[idx] = mrec
 
+			// If the deployment group is fulfilled then break out and
+			// move to the next deployment
 			if drec.Count == 0 {
-				continue mainloop
+				continue deploymentGroupLoop
```

Comment:
I'm sorry you had to read through this logic :joy: .  Thank you for annotating it.